k2/csrc/intersect.cu

```diff
@@ -90,950 +69,777 @@ using namespace intersect_internal;  // NOLINT
-    // boundary).
-    // the max will be -infinity for any FSA-id that doesn't have any active
-    // states (e.g. because that stream has finished).
-    // Casting to ragged2 just considers the top 2 indexes, ignoring the 3rd.
-    // i.e. it's indexed by [fsa_id][state].
-    Ragged<float> end_scores_per_fsa = arc_end_scores.RemoveAxis(1);
-    Array1<float> max_per_fsa(c_, end_scores_per_fsa.Dim0());
-    MaxPerSublist(end_scores_per_fsa, -std::numeric_limits<float>::infinity(),
-                  &max_per_fsa);
-    const int32_t *arc_end_scores_row_splits1_data =
-        arc_end_scores.RowSplits(1).Data();
-    const float *max_per_fsa_data = max_per_fsa.Data();
-    float *dynamic_beams_data = dynamic_beams_.Data();
-
-    float default_beam = search_beam_, max_active = max_active_,
-          min_active = min_active_;
-    K2_CHECK_LT(min_active, max_active);
-
-    Array1<float> cutoffs(c_, num_fsas);
-    float *cutoffs_data = cutoffs.Data();
-
-    auto lambda_set_beam_and_cutoffs =
-        [=] __host__ __device__(int32_t i) -> void {
-      float best_loglike = max_per_fsa_data[i],
-            dynamic_beam = dynamic_beams_data[i];
-      int32_t active_states = arc_end_scores_row_splits1_data[i + 1] -
-                              arc_end_scores_row_splits1_data[i];
-      if (active_states <= max_active) {
-        // Not constrained by max_active...
-        if (active_states >= min_active || active_states == 0) {
-          // Neither the max_active nor min_active constraints
-          // apply.  Gradually approach 'beam'
-          // (Also approach 'beam' if active_states == 0; we might as
-          // well, since there is nothing to prune here).
-          dynamic_beam = 0.8 * dynamic_beam + 0.2 * default_beam;
-        } else {
-          // We violated the min_active constraint -> increase beam
-          if (dynamic_beam < default_beam) dynamic_beam = default_beam;
-          // gradually make the beam larger as long
-          // as we are below min_active
-          dynamic_beam *= 1.25;
-        }
-      } else {
-        // We violated the max_active constraint -> decrease beam
-        if (dynamic_beam > default_beam) dynamic_beam = default_beam;
-        // Decrease the beam as long as we have more than
-        // max_active active states.
-        dynamic_beam *= 0.9;
+    // set up steps_, which contains a bunch of meta-information about the steps of the algorithm.
+    InitSteps();
+
+    int32_t num_seqs = b_fsas.shape.Dim0();
+
+    { // check that b_fsas are in order of decreasing length.
+      Array1<int32_t> r = b_fsas.shape.RowSplits(1).To(GetCpuContext());
+      int32_t *r_data = r.Data();
+      int32_t prev_t = r_data[1] - r_data[0];
+      for (int32_t i = 1; i + 1 < r.Dim(); i++) {
+        int32_t this_t = r_data[i+1] - r_data[i];
+        if (this_t < prev_t)
+          K2_LOG(FATAL) << "Sequences (DenseFsaVec) must be in sorted from greatest to least.";
+        prev_t = this_t;
```

Comment:
wondering if this is a check for only debug mode? why we do it on CPU?

Reply:
I figured that for typical number of FSAs, the overhead of transfer will probably exceed the time from doing it on CPU..